generators/chipyard-helpers/src/main/scala/chipyard-helpers/CodingStyle.scala

```diff
@@ -0,0 +1,53 @@
+class ExampleModule extends Module {
+  // val chisel stuff should be in snake case
+  val temp_wire = WireDefault(...)
+  val my_cool_reg = RegInit(...)
+
+  // scala global defaults should be all caps
+  val DEFAULT_VALUE = 20
+
+  // other scala vars should be in camelcase
+  val myOtherVar = 20
+
+  // scala functions that are associated with the module should be put in an obj
+  // notice how the function is also in camel case since it gives a scala value out
+  def myFoo(intBits: Int, chisel_val: UInt): Int = {
+
+  }
+
+  // notice how the function is also in snake case since it gives a chisel value out
+  def my_bar(intBits: Int, intBits2: Int): UInt = {
+
+  }
+}
+
+class ModuleTemplate {
+  // override this function to place all ios in a known location
+  // must be scaladoc commented
+  def ios: ??? = ???
+
+  // override this function to place all chisel code in a known location
+  // must be scaladoc commented
+  def implementation: ??? = ???
+}
+
+class ChiselModuleStart extends ModuleTemplate {
+  def multiplyScalaNumbers(inputNumber0: Int, inputNumber1: Int): Int = {
+    inputNumber0 * inputNumber1
+  }
+
+  def add_two_chisel_uints(in_0: UInt, in_1: UInt): UInt = {
+    in_0 + in_1
+  }
+
+  override def ios: ??? = {
+    val io = IO(new Bundle {
+      val my_io_here = Input(UInt())
+      val my_io_again_here = Output(UInt())
+    })
+  }
+
+  override def implementation: ??? = {
```

Comment:
I feel like this will just cause more confusion then it saves. some questions I foresee:

 - When should I use Module instead of ModuleTemplate?
 - Why does ModuleTemplate force this implementation block, while in the normal modules I can just put it in the class body?
 - What's the difference between a Module, a LazyModule, a LazyModuleImp, and a ModuleTemplate?
 - How do I mix ModuleTemplates and Modules?

Reply:
Comments or an accompanying document to brief explain the uses? At least the Lazymodule stuff. Don't know anything about ModuleTemplate

Reply:
> Comments or an accompanying document to brief explain the uses? At least the Lazymodule stuff. Don't know anything about ModuleTemplate

This type of template is not supposed to help people to understand `LazyModule` or `LazyModuleImp`'s. That we need to discuss more since that's significantly harder to explain concisely :). This work is attempting to do 2 things:

1. Have a coding style guide that others can reference when making new modules
2. Have a more restrictive form of a normal Chisel `Module` that (softly) forces users to organize code in a particular way (IOs in one place, the rest of the module in another place) so it is easier to read. I argue that if this isn't in a "template" form, users will just ignore this file and never adhere to it (unless someone wants to write a Chisel + Scala linter - which I think is the ultimate/best solution).

> I feel like this will just cause more confusion then it saves. some questions I foresee:
> 
> * When should I use Module instead of ModuleTemplate?

Yea, I think the optimal thing would be to somehow make this a trait that gets mixed in and forces an organization... but I don't know how that would work / if it could work.

> * Why does ModuleTemplate force this implementation block, while in the normal modules I can just put it in the class body?

Yea, I don't have a good response to this other than us writing good docs saying it's a very restrictive form of creating a module (for readability)

> * What's the difference between a Module, a LazyModule, a LazyModuleImp, and a ModuleTemplate?

We need to build up a story for the Lazy*'s and making them easier to understand.

> * How do I mix ModuleTemplates and Modules?

With the right documentation... you can say that they are interchangeable.

Reply:
My point is that this would introduce yet-another abstraction to module construction that would need to be explained, so we'd need to tread carefully here.
To help understand the problem more, can you provide examples of existing modules that would benefit from a more restrictive module format? (Benefit as in, improve readability)

Reply:
Points where I think this type of construct might help:

"Where are the IOs of this module?" - 
https://github.com/chipsalliance/rocket-chip/blob/47f7b7144727f0340d511d35b9f6c7a91b2a276f/src/main/scala/rocket/RocketCore.scala#L122

"What does it mean to have multiple IOs?" -
https://github.com/OpenXiangShan/XiangShan/blob/3eeae490776c53712da27da9143d26c0c8238996/src/main/scala/top/Top.scala#L116

"Are these functions utilities specific to the module or can they be useful elsewhere?" -
https://github.com/chipsalliance/rocket-chip/blob/47f7b7144727f0340d511d35b9f6c7a91b2a276f/src/main/scala/rocket/TLB.scala#L733

I think there is more... but stuff like this seems like a starting point. I think enforcing functions makes the boundaries clear of what should be put where (instead of allowing users to potentially put IOs anywhere, have multiple IO definitions, functions to be everywhere, etc)